utilities/io_tools/input_tools.py

```diff
@@ -279,35 +299,214 @@ def read_matched_scetlib_dyturbo_hist(scetlib_resum, scetlib_fo_sing, dyturbo_fo
     if "vars" in hfo.axes.name and hfo.axes["vars"].size != hfo_sing.axes["vars"].size:
         if hfo.axes["vars"].size == 1:
             hfo = hfo[{"vars" : 0}]
     hnonsing = hh.addHists(-1*hfo_sing, hfo, flow=False, by_ax_name=False)
+
     if fix_nons_bin0:
         # The 2 is for the WeightedSum
         res = np.zeros((*hnonsing[{"qT" : 0}].shape, 2))
         if "charge" in hnonsing.axes.name:
             hnonsing[...,0,:,:] = res
         else:
             hnonsing[...,0,:] = res
-    # TODO: Validate
-    if hnonsing.axes["vars"].size != hsing.axes["vars"].size:
-        htmp_nonsing = hsing.copy()
-        for var in htmp_nonsing.axes["vars"]:
-            logger.warning(f"Did not find variation {var} for nonsingular! Assuming nominal")
-            htmp_nonsing = hnonsing[{"vars" : var if var in hnonsing.axes["vars"] else 0}]
+
+    # variations are driven by resummed result, collect common variations from nonsingular piece
+    # if needed
+    if hnonsing.axes["vars"] != hresum.axes["vars"]:
+
+        # remapping is needed for scale variations which have slightly different parameter
+        # definitions for resummed vs fixed-order pieces
+        # *FIXME* this is sensitive to presence or absence of trailing zeros for kappas
+        scales_map = {"mufdown": "kappaf0.5",
+                      "mufup" : "kappaf2.",
+                      "mufdown-kappaFO0.5-kappaf2." : "kappaFO0.5",
+                      "mufup-kappaFO2.-kappaf0.5" : "kappaFO2.",
+                      }
+
+        htmp_nonsing = hist.Hist(*hnonsing.axes[:-1], hresum.axes["vars"], storage = hnonsing._storage_type())
+
+        for i, var in enumerate(hresum.axes["vars"]):
+            var_nonsing = scales_map.get(var, var)
+            if ("muf" in var or "kappaf" in var or "kappaFO" in var) and var_nonsing not in hnonsing.axes["vars"]:
+                raise ValueError(f"Scale variation {var} found for resummed piece which should correspond to {var_nonsing} for nonsingular piece but is not found")
+            var_nonsing = var_nonsing if var_nonsing in hnonsing.axes["vars"] else hnonsing.axes["vars"][0]
+
+            htmp_nonsing[{"vars" : i}] = hnonsing[{"vars" : var_nonsing}].view(flow=True)
+
         hnonsing = htmp_nonsing
-    return hh.addHists(hsing, hnonsing, by_ax_name=False)
+
+    htotal = hh.addHists(hresum, hnonsing, by_ax_name=False)
+
+    # remove duplicates and extend variations with some envelopes and special kinematic slices
+
+    vars_dict = {var : None for var in htotal.axes["vars"]}
+    vars_unique = vars_dict.keys()
+
+    vars_out = list(vars_unique)
+
+    additional_var_hists = {}
+
+    s = hist.tag.Slicer()
+
+    htotal_nominal =  htotal[{"vars" : "pdf0"}]
+
+    resum_scales = ["muB", "nuB", "muS", "nuS"]
+    resum_scale_vars = [var for var in vars_unique if any(resum_scale in var for resum_scale in resum_scales)]
+
+    if "kappaFO0.5-kappaf2." in vars_unique:
```

Comment:
Surely this code can be made generic enough to loop over the variations rather than listing them all explicitly. Regardless this should be a different function, it makes the function much more confusing to read

Reply:
Yes fair enough I'll clean this up a bit.